### ISSUE TITLE
[Feature Request] Plugin API, Phase V: The `__beartype_isinstance__()` protocol: Class-specific type-checking violation messages

### ISSUE THREAD

**Issue by leycec** [beartype/beartype#235]
This feature request comes courtesy @patrick-kidger (AKA, the Big Brain) at issue #216 – now split into smaller issues for easier digestion. :yum:

[Exception messages currently generated by @beartype **+** `jaxtyping` are appallingly unreadable](https://github.com/beartype/beartype/issues/216#issuecomment-1463308237):

```
beartype.roar.BeartypeCallHintParamViolation: @beartyped __main__.f() parameter
x="Array(1, dtype=int32, weak_type=True)" violates type hint <class
'jaxtyping.Float[Array, 'dim1']'>, as <protocol "jaxlib.xla_extension.ArrayImpl">
"Array(1, dtype=int32, weak_type=True)" not instance of <class
"jaxtyping.Float[Array, 'dim1']">.
```

:vomiting_face: 

I would punch the nearest pliable object repeatedly if I ever got an exception message like that. This is actually a desperate plea for a new variant of a @beartype plugin API in disguise. See also feature requests #191, #192, and #193. Of course, none of those do what you need. But they *do* demonstrate that a sprawling and poorly documented plugin system is a thing we should promise to do sometime.

I see two alternate APIs here:

* We run with the proposed `def __beartype_isinstance__(self, obj: object) -> tuple[bool, None | str]: ...` protocol. This works, but has a bit of a drawback: you're *probably* violating DRY by duplicating validation logic between this new `__beartype_isinstance__()` and the existing `__instancecheck__()` in the metaclass. Maybe? Though, I suppose that one could always just internally defer to the other *or* just drop `__instancecheck__()` entirely (in which case `jaxtyping` would strictly require users to use `@beartype` everywhere, which is just great from my biased perspective but not so great from everybody else's more honest perspective). Still, it seems a little excessive. I wonder if instead...
* We pare that down to an alternate `def __beartype_violation_message__(self, obj: object) -> None | str: ...` protocol, which returns either:
  * If the passed object `obj` satisfies the current type hint `self`, `None`.
  * Else, a human-readable substring detailing exactly how `obj` violates `self`.

In the latter case, you'd still define an `__instancecheck__()` in the metaclass; @beartype would then still defer to a standard `isinstance(obj, hint)` check like it currently does. If that check returns `False`, @beartype would then call your hint-specific `__beartype_violation_message__()` dunder method (if defined) to embed a hint-specific description in its violation message.

Since @beartype will embed that substring in its own larger exception message, there will *probably* need to be a few ground rules governing how exactly that message is formatted (e.g., as a lowercase sentence fragment with no prefixing or trailing punctuation). Blagh! Let's not even begin to consider l18n and i18n concerns here, because my feeble brain is already imploding.

:expressionless: :exploding_head: :boom:

**Comment by patrick-kidger** [comment]
@agronholm -- I've heard rumours (https://github.com/google/jaxtyping/issues/6) that typeguard is planning on adding support for a feature something like this as well.

Would it be possible for us all to standardise on a single API here?

**Comment by leycec** [comment]
:grimacing: 

I'm so sorry, @patrick-kidger. I wasn't able to nail this to the floor for `beartype 0.14.0`, which now needs to be shipped yesterday due to resolving a critical blocker for our sad-faced PyPy userbase. On a scale of "I don't even care, bro!" to "I will burn this issue tracker if this never happens!", how critical would you say this feature request is for `jaxtyping`?

I'm assuming super-critical, because that exception message is trash. Still... is this something you can live without for a bit? I'd kinda like to go hard on import hooks for `beartype 0.15.0` instead. I leave this decision in your ingenious hands that are hopefully less arthritic than mine.

**Comment by patrick-kidger** [comment]
Haha, it's totally fine -- not critical at all. Best of luck fixing those arthritic hands...

**Comment by leycec** [comment]
**Aww!** You're so noice – as always. Thanks so much for your *extreme* patience here and elsewhere with our swollen issue tracker that spawns ever more bugs. As the spring churns dutifully into summer, we reflect on the nature of exponential growth in insect populations on- and offline.

Now... I just gotta massage these bulging knuckles until the blood flow returns. Shouldn't that happen automatically, though!? :sweat:

**Comment by dkamm** [comment]
Hey @leycec - just wanted to check in on this. Would be nice to get those friendlier error messages

**Comment by leycec** [comment]
**Yes!** They're still awful, are they? `</sigh>`

I'm [hip-deep in import hooks](https://github.com/beartype/beartype/issues/43), which are such *massive hype* that everyone will be ecstatic that I did those instead of this. Following that, I'm afraid that [deep type-checking of `dict` and `set` collections](https://github.com/beartype/beartype/issues/167) is up next. @langfield must be appeased. He's also right; it's been nearly a year since he first justifiably complained about @beartype not type-checking `dict` and `set` items.

Relatedly, @beartype is just this fun hobby thing I do in my scarce free time in a purely volunteer manner mostly without any funding whatsoever. It is all for the :heart: here. Official funding would supercharge this development process, but financial utopia is hard to see from here. Until then, I happily flail my fat limbs like an open-source turtle flipped onto its back.

Sadly, that means `__beartype_isinstance__()` is currently on the back-burner until the Autumn. I'm still *absolutely* committed to this, because existing exception messages are trash for common third-party use cases like `jaxtyping`. But... I lack temporal bandwidth for this. Someone skilled in unethical bioengineering, clone me five hundred times! Let the Bear Clone Wars begin. :bear: :bear: :bear: :bear:

**Comment by leycec** [comment]
**Hah-ah!** I've been *sloooowly* speccing this API out with @tvdboom over at discussion #277. I'm now glad I hesitated to do anything. :grin: 

The resulting API is surprisingly non-trivial. Why? Recursion and subscription. For example, consider this `typing.Protocol` subclass defined by @tvdboom over at #277:

```python
from beartype.typing import Protocol, TypeVar

T = TypeVar('T')
class SequenceProtocol(Protocol[T]):
    def __iter__(self) -> Iterable[T]: ...
    def __getitem__(self, item) -> T: ...
    def __len__(self) -> int: ...
```

`SequenceProtocol` needs to be able to generate human-readable violation messages in all possible use cases – including when:

* Unsubscripted (e.g., as the vanilla `SequenceProtocol` class).
* Subscripted by a non-nested child type hint (e.g., as `SequenceProtocol[int]`).
* Subscripted by a nested instance of itself (e.g., as `SequenceProtocol[SequenceProtocol[str]]`).

Here's where our [existing object-oriented `beartype.door.TypeHint` API](https://beartype.readthedocs.io/en/latest/api_door/#object-oriented-api) comes in. @beartype's `TypeHint` is already *almost* smart enough to support the above use cases. We just need a bit of extra special sauce. What follows is mostly myself talking to myself, because I *will* forget this in five minutes after mucking about in the garden with smelly cats:

* Define a new private `beartype.door._doorerror` submodule.
* Publicize our currently private `beartype._decor.error._errorcause.ViolationCause` dataclass as a public `beartype.door.TypeHintCause` dataclass privately implemented at `beartype.door._doorerror.TypeHintCause`.

`beartype.door.TypeHintCause` is a new public dataclass enabling third-party frameworks to control the generation of framework-specific violation messages. In the case of @tvdboom's `SequenceProtocol`, the final implementation from the end user perspective would resemble:

```python
from beartype.door import TypeHint, TypeHintCause
from beartype.typing import Protocol, TypeVar

T = TypeVar('T')
class SequenceProtocol(Protocol[T]):
    def __iter__(self) -> Iterable[T]: ...
    def __getitem__(self, item) -> T: ...
    def __len__(self) -> int: ...

    # FUN STUFF FOLLOWS. START READING HERE, HOPEFULLY FELLOW HUMAN.
    def __beartype_cause__(cls, cause: TypeHintCause) -> TypeHintCause:
        '''
        Output cause describing whether the object currently being type-checked
        by the passed input cause either satisfies or violates this protocol.

        Parameters
        ----------
        cause : ViolationCause
            Input cause describing the object currently being type-checked.

        Returns
        ----------
        ViolationCause
            Output cause refining this description.
        '''

        # If the object currently being type-checked is *NOT* a sequence,
        # define an appropriate error message and immediately return.
        if not isinstance(cause.obj, cls):
            cause.message = 'not a sequence'
        # Else, this object is a sequence.
        #
        # If this sequence is non-empty...
        elif cause.obj:
            # First child hint of this hint. All remaining child hints if any
            # are ignorable. Specifically, if this hint is:
            # * A standard sequence (e.g., "typing.List[str]"), this hint is
            #   subscripted by only one child hint.
            # * A variadic tuple (e.g., "typing.Tuple[str, ...]"), this hint is
            #   subscripted by only two child hints the latter of which is
            #   ignorable syntactic chuff.
            item_hint: TypeHint = cause.hint[0]

            # If this child hint is *NOT* ignorable...
            if not item_hint.is_hint_ignorable:
                # For each enumerated item of this sequence...
                for item_index, item in enumerate(cause.obj):
                    # Deep output cause, type-checking whether this item
                    # satisfies this child hint.
                    item_cause = cause.permute(
                        obj=item, hint=item_hint).refine()

                    # If this item is the cause of this failure...
                    if item_cause.message is not None:
                        # Human-readable substring prefixing this failure with
                        # metadata describing this item.
                        item_cause.message = (
                            f'{color_type(label_object_type(cause.obj))} '
                            f'index {item_index} item {item_cause.message}'
                        )

                        # Return this cause.
                        return item_cause
                    # Else, this item is *NOT* the cause of this failure.
                    # Silently continue to the next item.
            # Else, this child hint is ignorable.
        # Else, this sequence is empty, in which case all items of this sequence
        # (of which there are none) are valid. Just go with it, people.

        # Return this cause as is; all items of this sequence are valid,
        # implying this sequence deeply satisfies this hint.
        return cause
```

`__beartype_cause__()` looks hairy, but that's just because I'm verbose (read: I like the sound of my own voice). In actuality, that's a worst-case `__beartype_cause__()` implementation – and it's still only ~20 lines of code or so to recursively generate human-readable violation messages for arbitrarily complex and possibly nested sequences.

Moreover, that `__beartype_cause__()` implementation is copied almost verbatim from our existing `beartype._decor.error._pep._pep484585._errorsequence._find_cause_sequence()` function. All we're *really* doing here is publicizing private stuff that already exists and has been known to work for years in the intestinal walls of the @beartype codebase.

**Comment by tvdboom** [comment]
This would be awesome! Can't wait for the implementation.

But how would this look like for somethign way simpler, where we don't need a custom protocol class? for example

```python
Annotated[int, Is[lambda v: v>0]]
```

I want this to error with something like: Value should be larger than zero, got X.

**Comment by patrick-kidger** [comment]
> <leycec's post>

Hang on, this API looks odd to me.

- Why is the `cause` passing information about the hint? That's not needed at all, the hint already knows who it is via the `cls` parameter.
- The mutation in `cause.message = ...` doesn't seem necessary.
- This seems to require learning a whole new system for the end user. I think that's a non-starter from a usability perspective.

I'm just imagining something like this:
```python
from typing import Generic, Optional, TypeVar, get_args, get_origin

### library

def _default_beartype_instancecheck(cls, item) -> Optional[str]:
    if isinstance(item, cls):  # in practice also handle non-class hints here
        return None
    else:
        return f"{item} not of type {cls}"

class Message:
    __slots__ = ("msg",)
    def __init__(self, msg: Optional[str]):
        self.msg = msg

    def __bool__(self):
        return self.msg is None

    def __str__(self):
        return str(self.msg)

def beartype_isinstance(item, hint) -> Message:
    cls = get_origin(hint)
    if cls is None:
        cls = hint
    if hasattr(type(cls), "__beartype_instancecheck__"):
        checker = type(cls).__beartype_instancecheck__
    else:
        checker = _default_beartype_instancecheck
    out = checker(hint, item)
    # Return in something that has the right `__bool__`, so that
    # `if beartype_isinstance(...):` has the same semantics as `if isinstance(...):`.
    return Message(out)

### user code

T = TypeVar("T")

class MetaA(type):
    def __beartype_instancecheck__(cls, item):
        (typevar,) = get_args(cls)
        if beartype_isinstance(item, typevar):  # recurse properly, friends
            return None
        else:
            return f"Wahoo, a custom error message. You didn't pass a {cls}!"

class LooksLike(Generic[T], metaclass=MetaA):
    pass

print(beartype_isinstance(1, int))  # None
print(beartype_isinstance(1, str))  # 1 not of type <class 'str'>
print(beartype_isinstance(1, LooksLike[int]))  # None
print(beartype_isinstance(1, LooksLike[str]))  # Wahoo, a custom error message. You didn't pass a __main__.LooksLike[str]!
```

**Comment by leycec** [comment]
@tvdboom: **Sweetness.** Simplicity is my middle name! It's not, actually. But... *it should have been.*

The above implementation of your `SequenceProtocol.__beartype_cause__()` method is definitely the worst case; that method necessarily supports deeply nested child type hints (e.g., `SequenceProtocol[Dict[str, SequenceProtocol[str]]` sort of stuff). Ideally, your newest example of `Annotated[int, Is[lambda v: v>0]]` would be the best case. That type hint can't possibly be subscripted by anything, so... it should be trivial to define an `__beartype_cause__()` method describing that type hint. Right?

Well, sorta. If @beartype allowed you to pass a custom `cause=` keyword parameter inside subscriptions of the `Is[...]` validator, then the following would suffice – maybe, probably, hopefully, *ohgodsisthisevergonnawork*:

```python
from beartype.door import TypeHintCause
from beartype.typing import Annotated
from beartype.vale import Is

def positive_int_cause(cause: TypeHintCause) -> TypeHintCause:
    '''
    ``__beartype_cause__()`` implementation for the :pep:`593`-compliant
    :data:`.PositiveInt` type hint declared below.
    '''

    # If the object currently being type-checked is *NOT* an integer, define an
    # appropriate error message.
    if not isinstance(cause.obj, int):
        cause.message = 'not an integer'
    # Else, this object is an integer.
    #
    # If this integer is non-positive, define an appropriate error message.
    elif cause.obj <= 0:
        cause.message = f'value should be larger than zero, got {cause.obj}'
    # Else, this integer is positive and thus satisfies "PositiveInt". Notify
    # the caller of this by doing absolutely nothing. Just go with it, people.

    # Return this cause.
    return cause

PositiveInt = Annotated[int, Is[lambda v: v>0, cause=positive_int_cause]]
'''
:pep:`593`-compliant type hint matching a positive integer.
'''
```

**Super simple.** That's just 7 additional lines of code, ignoring extraneous docstrings and comments. Not needing to support child type hints alleviates all of the anguish associated with the above `SequenceProtocol.__beartype_cause__()` implementation. This is why we rejoice. :partying_face:

**Comment by leycec** [comment]
> Hang on, this API looks odd to me.

*...heh.* This API looks odd to you, because this API is odd. So, let's back up. The core @beartype decorator dynamically generates Python code that does two sequential things to type-check type hints at runtime. In order, these things are:

1. **Testing.** First, @beartype *efficiently* tests whether the parameters passed to and value returned from the current callable satisfy the type hints annotating that callable. @beartype works *very* hard to ensure that this test is both macro- and micro-optimized to the hilt *and* that this test guaranteeably runs in `O(1)` time with negligible constant factors.
2. **Exception raising.** Second, if one or more of these tests fail, @beartype then *inefficiently* raises a human-readable exception. This phase is completely unoptimized. It's expected to be slow and typically is slow, but nobody cares because it hardly ever happens. If something failed a type-check, then your code is busted and is about to crash anyway. Efficiency is no longer a pressing concern. You have bigger problems.

One of the main philosophical realizations of @beartype is that runtime type-checking can be done *extremely* fast if you separate the testing and exception-raising parts of runtime type-checking into two distinct phases. Testing can be done fast – but only so long as all you're doing is *testing*. As soon as you try actually generating a human-readable exception message, you're now slow. Thus the separation. Ideally, correctly functioning code will do a *lot* of runtime type-checks that *always* pass these fast tests and thus never even need to go on to the slow portion that raises a human-readable exception on a failing type-check.

So far, so good. Here's where things get less good. Surprisingly, raising human-readable exceptions is non-trivial – especially when you consider non-trivial parent type hints that arbitrarily nest child type hints – especially when some of those child type hints imply conditional branching logic. Something like `Union[List[str], MutableMapping[int, Iterable[Collection[Sequence[int]]]]` is a horrifying example.

This weirdo API that @beartype will soon be exposing to you? Yeah. That weirdo API has been battle-hardened to handle just these kinds of monstrous type hints that nobody should ever define but (contrary to sanity) everybody defines all the time.

This weirdo API is currently hidden from you in the `beartype._decor.error` subpackage, which contains almost ~2,500 lines of code alone *just to generate human-readable type-checking violation messages*. This is why you are squinting. Because this weirdo API is already capable of doing things far in excess of what the standard use case requires. But... there's really *no* other way to do this. Thankfully, as the *much* conciser implementation of the `positive_int_cause()` function above demonstrates, this weirdo API is *much* simpler than the initial worst-case `SequnceProtocol.__beartype_cause()` method above made it out to be.

It's just when you want to do hard things like handle child type hints that things start spiralling out of control.

*...oh, gods.* What are we even doing here!?!?  :face_exhaling: 

> Why is the cause passing information about the hint? That's not needed at all, the hint already knows who it is via the `cls` parameter.

*...heh.* That's actually needed. The `cls` parameter fails to convey subscriptions. `typing.Protocol` subclasses `typing.Generic`. Think `def SequenceProtocol(Protocol[T]):` in the above example. That's a subscriptable generic – not simply a vanilla unsubscriptable class. All PEP 544-compliant protocols are also PEP 484-compliant generics and can be subscripted as such.

Indeed, @tvdboom's use case *requires* subscription. Consider `SequenceProtocol[int]`, which effectively carries the same typing semantics as `collections.abc.Sequence[int]` with the added benefit of supporting non-standard sequences that satisfy `SequenceProtocol` but *not* `collections.abc.Sequence`. These include (*...waitforit*) NumPy and thus JAX arrays as well as Pandas dataframes, indices, and series.

It's actually a pretty cool idea. I personally applaud @tvdboom for this! Publish `SequenceProtocol` as an open-source package of some sort and reap the incoming GitHub stars, partying emoji, and positive life karma. This is the way.

> The mutation in `cause.message = ...` doesn't seem necessary.

*...heh.* You'd prefer to just return a simple dataclass like `Message` encapsulating a string message, right? I certainly get it. I myself prefer that. That's the standard approach and what everyone would prefer to happen rather than this convoluted dataclass abstraction called `TypeHintCause` that you then mutate, permute, and return all over the place.

Sadly, the dataclass is *probably* necessary. Why? Because the `TypeHintCause` API does *alot* more than just generate messages. `TypeHintCause` also does essential bookkeeping chores like identifying the [exact culprit responsible for the current type-checking violation](https://beartype.readthedocs.io/en/latest/api_roar/#beartype.roar.BeartypeCallHintViolation.culprits), where "culprit" means the `beartype.roar.BeartypeCallHintViolation.culprits` instance variable associated with all type-checking violations.

Culprits is kinda a deep topic. But the point is that the `TypeHintCause` API also needs to identify the exact object responsible for a type-checking violation. This is trivial in the case of a simple type, as the above `PositiveInt` example demonstrates. But this is non-trivial in the case of a subscriptable type hint, as the above `SequenceProtocol` example demonstrates. In the latter case, the culprit is typically an item of an iterable container. That item needs to be:

* **Identified** (via iteration in `SequenceProtocol.__beartype_cause__()`).
* **Exposed** (via the `BeartypeCallHintViolation.culprits` variable).

Sunlight is the best disinfectant for the darkness. That's a great phrase. Thanks, dead Supreme Court Justice Louis Brandeis! But will anyone actually want to do any of this? I... don't know. @beartype already does this. So, at least one living human wants to do this. Yep. It's me. I sigh and then grimace. :face_exhaling: :grimacing: 

You are now thinking to yourself: "But... I don't *care* about culprits or any of that other stuff. I just care about exception messages. Can't we just focus on that, you cretin?" I know you're thinking that, Dr. Kidger. I know you. You demand sane APIs. Admittedly, I am a cretin.

But... @beartype is kinda insane. The problem domain of runtime type-checking is ostensibly a [wicked problem](https://en.wikipedia.org/wiki/Wicked_problem). You want to just care about exception messages – but that's not really possible, because generating exception messages is inextricably entangled with identifying culprits and all of that other bookkeeping. Philosophically and programmatically, these problems are all interrelated.

Consider, for example: how do you even generate a human-readable exception message without first identifying the culprits responsible for the current type-checking violation? You can't, really. First, you figure out which object committed seppuku; this is the culprit. Then, you shame that culprit by describing that culprit in an exception message.
 
> This seems to require learning a whole new system for the end user.

**Yup.**

> I think that's a non-starter from a usability perspective.

**Yup.** I'm sorta fine with that, because this is what @beartype itself internally needs to both generate human-readable exceptions as well as identify culprits for arbitrarily complex type hints. Those things were never going to be easy. They can't be. Indeed, those things are almost harder than dynamically generating code to efficiently detect type-checking violations.

And Odin only knows: *that's hard.*

Also...

```python
from typing import ..., get_args, get_origin
```

**Beware!** Pretty much *all* runtime `typing` functionality is busted. This includes both `typing.get_args()` and `typing.get_origin()`. None of those functions should ever be called in production-ready code, because none of those functions safely accept arbitrary type hints or behave as expected in various common edge cases.

Instead, this is why the `beartype.door.TypeHint` API exists – and why the `beartype.door.TypeHintCause` dataclass internally encapsulates the current type hint as the `cause.hint: TypeHint` instance variable. You *definitely* want to use that instead of blatantly busted `typing` functions.

Fascinatingly, your custom `_default_beartype_instancecheck()` and `beartype_isinstance()` functions and custom `Message` dataclass closely resemble my initial implementations in the `beartype._decor.error` subpackage. Naturally, things got explosively more complicated as @beartype began supporting more and more PEPs and various kinds of type hints. Memories of those early days are still painful.

Even something that you would *think* is simple like `typing.Union[...]` becomes extremely non-simple when you consider that @beartype needs to explain why exactly each child type hint subscripting the union fails to satisfy that union – both in human-readable form via the exception message *and* in machine-readable form via the `culprits` variable. Now, a single `Message` no longer suffices. You need one `Message` for each branch of the union describing exactly why that branch failed to satisfy its child type hint.

Your `Message` dataclass very rapidly begins to resemble our `TypeHintCause` dataclass – with all of its attendant API complexities, nuances, and usability migraines. Trust me, bro. This is the only way.

Never trust anyone who says:

> Trust me, bro.

**Comment by patrick-kidger** [comment]
Separating testing and raising is a good point, and I like that.

> The cls parameter fails to convey subscriptions.

Look again! My implementation includes passing the subscripted hint as the `cls`. IMO this is cleaner, as it's just a subscripted generic, not a new `TypeHint` object.

> The mutation in cause.message = ... doesn't seem necessary.

What I'm saying is, you're already returning `cause`. You don't need to mutate an input as well. An API should typically either mutate its input, or (preferably) return a brand-new object. Not both.

As you point out, this is already on the slow path, so there's no interesting speed advantage to mutation.

(The fact that it's a dataclass is totally fine.)

> This seems to require learning a whole new system for the end user.

Do I understand correctly that the reason for this proposed system, as opposed to just "return a string", is specifically (and only) so that implementers of `__beartype_cause__` are able to identify a culprit other than just `cls`? They might want to say that it's actually the fault of
```
cls[foo, bar[Union[int, str]]]
                   ^^^
                   this bit
```
?

> Pretty much all runtime typing functionality is busted. This includes both typing.get_args() and typing.get_origin()

Haha, I'm definitely willing to believe that. What's the failure mode for these two functions?

**Comment by leycec** [comment]
**OMG.** I... totally neglected to reply to @patrick-kidger. I'm sweating in a poorly insulated cabin over here. :sweat_smile:

On the bright side, @beartype is finally doing this. Specifically, our upcoming @beartype 0.17.0 release now supports our first official plugin API. And, no; it's *not* that hideous `__beartype_cause__()` thing that everybody hates. Instead, it's a @beartype-agnostic `__instancecheck_str__()` dunder method that everybody had better love. Like `__instancecheck__()`, `__instancecheck_str__()` is intended to be defined on the metaclass of user-defined classes. It has a signature resembling:

```python
def __instancecheck_str__(cls, obj: typing.Any) -> str:
```

Simple, right? And it actually is. `__instancecheck_str__()` is intended to be supported by competing runtime type-checkers (e.g., `typeguard`, Pydantic) as a pseudo-standard. Ain't nobody got the time to write an actual PEP for that, of course. Thus, pseudo-standard.

@patrick-kidger's killer `jaxtyping` will probably be the first third-party consumer. Let's *goooooooo*! It's a plugin party and you're all invited.

**Comment by leycec** [comment]
**Resolved by ef26913b4ac58d71f.** Thanks for your patience, for*bear*ance, and... *wait.* Was that an unexpected (but altogether welcome) bear pun in my @beartype feature request? You're not wrong. Where's there's puns, there's fire. This is your final API for raising human-readable violations:

```python
from beartype import beartype

class MetaclassOfMuhClass(type):
    def __instancecheck_str__(cls, obj: object) -> str:
        return f'{repr(obj)} has disappointed {repr(cls)}... for the last time.'

class MuhClass(object, metaclass=MetaclassOfMuhClass):
    pass

@beartype
def muh_func(muh_obj: MuhClass) -> None:
    pass

muh_func("Some strings, you just can't reach.")
```

...which now raises the "human"-readable violation message:

```python
beartype.roar.BeartypeCallHintParamViolation: Function __main__.muh_func() parameter
muh_obj="Some strings, you just can't reach." violates type hint <class '__main__.MuhClass'>, as
"Some strings, you just can't reach." has disappointed <class '__main__.MuhClass'>... for the last time.
```

Note that the string you return from your `__instancecheck_str__()` implementation:

* Should be a sentence fragment whose first character is typically *not* capitalized.
* May optionally be suffixed by a trailing period. In either case, @beartype intelligently does the right thing and ensures that the entire violation message is *always* suffixed by only a single trailing period.

Bear muscles are glistening with sweat. Flex it, @beartype! :muscle: :bear:
